engine/src/main/scala/cromwell/engine/workflow/lifecycle/materialization/MaterializeWorkflowDescriptorActor.scala

```diff
@@ -180,18 +184,36 @@ object MaterializeWorkflowDescriptorActor {
         s"'$optionName' is specified in workflow options but value is not of expected Double type: ${e.getMessage}".invalidNel
     }
   }
+
+  def getImportAuthProviders(conf: Config,
+                             authProviderFunc: String => ImportAuthProvider
+  ): ErrorOr[List[ImportAuthProvider]] = {
+    val isPrivateWorkflowsEnabled = conf.as[Boolean]("private-workflows.enabled")
+    val isAuthAzure = if (conf.hasPath("private-workflows.auth.azure")) {
+      conf.as[Boolean]("private-workflows.auth.azure")
+    } else false
+
+    if (isPrivateWorkflowsEnabled && isAuthAzure) {
+      val azureToken = AzureCredentials.getAccessToken()
```

Comment:
I am actually not sure where that WDL caching happens. 🤔 

>then often not using it

also just to note, it is possible we don't end up using the token even when the workflow is a public workflow.

Reply:
So for Azure, we generate this token once for every workflow we run, except potentially cached ones? Do you know whether it happens for each subworkflow as well?

It can take a while (like a couple seconds) to generate this token, I don't know that it's required for this PR but I think we'll eventually want some caching. It would be good to confirm whether we're already benefiting from the WDL caching, just so we understand the behavior. If we are, I think this is fine as-is.

Reply:
One thing to be wary of is that private workflows only get cached for the same user. Otherwise we have a potential data leak situation possible

Reply:
It seems that WDL caching is happening after the workflow url is resolved in the MaterializeDescriptorActor. 
- Workflow source/url resolution happens [here](https://github.com/broadinstitute/cromwell/blob/21e8180c2118ec8fd8c45d4c5f98aeb77dff46f5/engine/src/main/scala/cromwell/engine/workflow/lifecycle/materialization/MaterializeWorkflowDescriptorActor.scala#L373-L379)
- once it is resolved it calls [buildValidatedNamespace(...) here](https://github.com/broadinstitute/cromwell/blob/21e8180c2118ec8fd8c45d4c5f98aeb77dff46f5/engine/src/main/scala/cromwell/engine/workflow/lifecycle/materialization/MaterializeWorkflowDescriptorActor.scala#L383)
- `buildValidatedNamespace` calls `validateNamespace` and each WDL language has its own implementation. Inside `validateNamespace` it calls [retrieveOrCalculate](https://github.com/broadinstitute/cromwell/blob/21e8180c2118ec8fd8c45d4c5f98aeb77dff46f5/languageFactories/language-factory-core/src/main/scala/cromwell/languages/util/ParserCache.scala#L25) which does the WDL caching

So it seems that upon running 100 same workflows in a single Cromwell Runner in Azure, it would fetch Azure token each time to resolve the workflow url and then the WDL caching will happening during WOM generation.